### PR TITLE
Refactor how signatures/trampolines are stored in `Store`

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit refactors where trampolines and signature information is
stored within a `Store`, namely moving them from
`wasmtime_runtime::Instance` instead to `Store` itself. The goal here is
to remove an allocation inside of an `Instance` and make them a bit
cheaper to create. Additionally this should open up future possibilities
like not creating duplicate trampolines for signatures already in the
`Store` when using `Func::new`.

### CODE REVIEW

File: crates/wasmtime/src/sig_registry.rs [hunk 1]
```diff
@@ -14,20 +13,23 @@ use wasmtime_environ::{ir, wasm::WasmFuncType};
 #[derive(Debug, Default)]
 pub struct SignatureRegistry {
     wasm2index: HashMap<WasmFuncType, VMSharedSignatureIndex>,
+    index_map: HashMap<VMSharedSignatureIndex, Entry>,
+}
```
Comment (pepyakin) [inline crates/wasmtime/src/sig_registry.rs:17]: Since this now stores trampolines, it would be great to update the comment.

Comment (alexcrichton) [reply crates/wasmtime/src/sig_registry.rs:22]: Indeed!

**Comment by alexcrichton** [comment]
@fitzgen would you be able to take a look at this?

File: crates/wasmtime/src/sig_registry.rs [hunk 2]
```diff
@@ -36,15 +49,18 @@ impl SignatureRegistry {
...
-                    std::u32::MAX as usize,
+                assert!(
+                    len < std::u32::MAX as usize,
                     "Invariant check: signature_hash.len() < std::u32::MAX"
                 );
+                debug_assert_eq!(len, self.index_map.len());
                 let index = VMSharedSignatureIndex::new(u32::try_from(len).unwrap());
+                self.index_map.push(Entry {
+                    wasm: wasm.clone(),
+                    native: native.clone(),
```
Comment (fitzgen) [inline crates/wasmtime/src/sig_registry.rs:60]: Since this function has to take ownership, we might as well make the caller do the clone (if necessary) so that we can avoid clones when they aren't needed. That is, revert the change from taking `&WasmFuncType` and `&ir::Signature` back to taking `WasmFuncType` and `ir::Signature`.

**Review by fitzgen — APPROVED** [review]

File: crates/wasmtime/src/sig_registry.rs (hunk 2 above)
Comment (alexcrichton) [reply crates/wasmtime/src/sig_registry.rs:60]: This is sort of the `Entry` problem where it's problematic that getting an `entry` requires `wasm.clone()` (should probably just so a `get` followed by an `insert`, but also that the `native` isn't always inserted here (although it sometimes is)

If we think this is an issue I could change the arguments to `impl Into<Cow<'_, ...>>`, but I was hoping to optimize more for the cache hit use case here rather than the "need to insert case". Do you think I should try to go the `Cow` route though?

Comment (fitzgen) [reply crates/wasmtime/src/sig_registry.rs:60]: If it doesn't require bending over backwards too much, it would be nice.

Comment (alexcrichton) [reply crates/wasmtime/src/sig_registry.rs:60]: Hm it turns out that `Into<Cow>` doesn't have as many implementations as I thought it did so I don't think this will work easily. Digging in more though I'm realizing that we're doing a number of unnecessary allocations in a number of places. In general a `Module` has no need to store the native signature types (it only needs the wasm signatures).

Would you be ok landing this as-is and I can have a follow-up which I think makes all the callers only have `&WasmFuncType` anyway so there's no gain to be had by passing in an owned value?

Comment (fitzgen) [reply crates/wasmtime/src/sig_registry.rs:60]: Yeah, for sure
